File: VB/CustomServices/CustomServiceWithKeyUserControl.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Utils.MVVM
Imports DevExpress.XtraEditors
Imports System
Imports System.Windows.Forms

Namespace DxSampleCustomServices
    Partial Public Class CustomServiceWithKeyUserControl
        Inherits DevExpress.XtraEditors.XtraUserControl
        Public Sub New() 
            InitializeComponent()
            Dim mvvmContext As New MVVMContext()
            mvvmContext.ContainerControl = Me
        
            Dim commandButton As New SimpleButton()
            commandButton.Text = "Execute Command"
            commandButton.Dock = DockStyle.Top
            commandButton.Parent = Me

            mvvmContext.ViewModelType = GetType(ViewModelWithKeyedCustomService)
            ' Custom service registration
            mvvmContext.RegisterService("SomeKey1", New CustomService("Custom Service 1"))
            mvvmContext.RegisterService("SomeKey2", New CustomService("Custom Service 2"))
            ' UI binding for button
            mvvmContext.BindCommand(Of ViewModelWithKeyedCustomService)(commandButton, Sub(x) x.DoSomethingViaCustomServices())
        End Sub
         Private Sub OnDisposing()
            Dim context = MVVMContext.FromControl(Me)
            If context IsNot Nothing Then
                context.Dispose()
            End If
        End Sub
        Public Class ViewModelWithKeyedCustomService
            <DevExpress.Mvvm.DataAnnotations.ServiceProperty(Key:="SomeKey1")>
            Protected Overridable ReadOnly Property CustomService1() As ICustomService
                Get
                    Throw New NotImplementedException()
                End Get
            End Property
            <DevExpress.Mvvm.DataAnnotations.ServiceProperty(Key:="SomeKey2")>
            Protected Overridable ReadOnly Property CustomService2() As ICustomService
                Get
                    Throw New NotImplementedException()
                End Get
            End Property
            Public Sub DoSomethingViaCustomServices()
                CustomService1.DoSomething()
                CustomService2.DoSomething()
            End Sub
        End Class
        Public Interface ICustomService
            Sub DoSomething()
        End Interface
        Public Class CustomService
            Implements ICustomService
        
            Private text As String
            Public Sub New(Optional ByVal text As String = Nothing)
                Me.text = text
            End Sub
            Public Sub DoSomething() Implements ICustomService.DoSomething
                XtraMessageBox.Show("Hi!", If(text, "CustomService"))
            End Sub
        End Class
    	End Class
End Namespace

File: CS/CustomServices/CustomServiceWithKeyUserControl.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Utils.MVVM;
using DevExpress.XtraEditors;
using System;
using System.Windows.Forms;

namespace DxSampleCustomServices {
    public partial class CustomServiceWithKeyUserControl : XtraUserControl {
        public CustomServiceWithKeyUserControl() {
            InitializeComponent();
            #region SetUp
            MVVMContext mvvmContext = new MVVMContext();
            mvvmContext.ContainerControl = this;
        
            SimpleButton commandButton = new SimpleButton();
            commandButton.Text = "Execute Command";
            commandButton.Dock = DockStyle.Top;
            commandButton.Parent = this;
            #endregion SetUp

            #region #customServiceWithKey
            mvvmContext.ViewModelType = typeof(ViewModelWithKeyedCustomService);
            // Custom service registration
            mvvmContext.RegisterService("SomeKey1", new CustomService("Custom Service 1"));
            mvvmContext.RegisterService("SomeKey2", new CustomService("Custom Service 2"));
            // UI binding for button
            mvvmContext.BindCommand<ViewModelWithKeyedCustomService>(commandButton, x => x.DoSomethingViaCustomServices());
            #endregion #customServiceWithKey
        }
        #region CleanUp
        void OnDisposing() {
            var context = MVVMContext.FromControl(this);
            if(context != null) context.Dispose();
        }
        #endregion CleanUp
        public class ViewModelWithKeyedCustomService {
            [DevExpress.Mvvm.DataAnnotations.ServiceProperty(Key = "SomeKey1")]
            protected virtual ICustomService CustomService1 {
                get { throw new NotImplementedException(); }
            }
            [DevExpress.Mvvm.DataAnnotations.ServiceProperty(Key = "SomeKey2")]
            protected virtual ICustomService CustomService2 {
                get { throw new NotImplementedException(); }
            }
            public void DoSomethingViaCustomServices() {
                CustomService1.DoSomething();
                CustomService2.DoSomething();
            }
        }
        public interface ICustomService {
            void DoSomething();
        }
        public class CustomService : ICustomService {
            string text;
            public CustomService(string text = null) {
                this.text = text;
            }
            public void DoSomething() {
                XtraMessageBox.Show("Hi!", text ?? "CustomService");
            }
        }
    }
}